### PR TITLE
Feature. Investment contract

### PR DESCRIPTION
**Pull request by niowa**

### CODE REVIEW

File: test/Crowdsale.test.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+const Token = artifacts.require('Token.sol');
+const Crowdsale = artifacts.require('Crowdsale.sol');
+
+const tokenCost = 100;
+const rate = 2;
+const newRate = 3;
+const newTokenCost = 200;
+const transferAmount = 100;
+const notEnoughAmount = 50;
```
Comment (EugeneBiazrukau) [inline test/Crowdsale.test.js]: `transferAmount`, `investorsBalances and`, `notEnoughAmount` are not used

File: contracts/Interfaces/CrowdsaleInterface.sol [hunk 2]
```diff
@@ -0,0 +1,9 @@
+pragma solidity 0.4.24;
+
+import "./PlayChipTokenInterface.sol";
+
+/// @title Interface of ownable Public sale which can transfer tokens for investments from crowdsale and wallet app
+contract CrowdsaleInterface {
```
Comment (EugeneBiazrukau) [inline contracts/Interfaces/CrowdsaleInterface.sol]: Where are you going to use this interface?

File: contracts/Interfaces/CrowdsaleInterface.sol [hunk 3]
```diff
@@ -0,0 +1,9 @@
+pragma solidity 0.4.24;
+
+import "./PlayChipTokenInterface.sol";
+
+/// @title Interface of ownable Public sale which can transfer tokens for investments from crowdsale and wallet app
```
Comment (EugeneBiazrukau) [inline contracts/Interfaces/CrowdsaleInterface.sol]: Please add two blank lines, solhint rule

File: contracts/Interfaces/PlayChipTokenInterface.sol [hunk 4]
```diff
@@ -3,5 +3,6 @@ pragma solidity 0.4.24;
 import "../Ownable.sol";
 import "./ERC223.sol";
 
-
-contract PlayChipTokenInterface is ERC223, Ownable {}
+contract PlayChipTokenInterface is ERC223, Ownable {
```
Comment (EugeneBiazrukau) [inline contracts/Interfaces/PlayChipTokenInterface.sol]: ` Two blank lines rule` is working here too

File: contracts/Crowdsale.sol [hunk 5]
```diff
@@ -0,0 +1,67 @@
+pragma solidity 0.4.24;
+
+import "./Ownable.sol";
+import "./SafeMath.sol";
+import "./Interfaces/PlayChipTokenInterface.sol";
+import "./Interfaces/ERC223RecieverInterface.sol";
+
+/// @title Crowdsale with could sell tokens for eth
```
Comment (EugeneBiazrukau) [inline contracts/Crowdsale.sol]: Two blank lines rule is working here too

File: contracts/Crowdsale.sol [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    emit FundTransfer(msg.sender, amount); // solhint-disable-line
+    withdrawAddress.transfer(amount);
+  }
+
+  function setRate(uint _rate) public onlyOwner {
+    rate = _rate;
+  }
+
+  function convertEthToTokens(uint _amount) public constant returns (uint convertedAmount) {
```
Comment (EugeneBiazrukau) [inline contracts/Crowdsale.sol]: Please, use `view`modifier.  `Constant`modifier will be dropped at 0.5.0

File: contracts/Token.sol [hunk 7]
```diff
@@ -35,16 +36,18 @@ contract Token is Ownable, ERC223, SafeMath {
     msg.sender.transfer(_amount);
   }
 
+  function setTokenGenerator(address _tokeGenerator) external {
+    tokenGenerator = _tokeGenerator;
```
Comment (EugeneBiazrukau) [inline contracts/Token.sol]: Please, add check for non-zero address

File: contracts/Token.sol [hunk 8]
```diff
@@ -35,16 +36,18 @@ contract Token is Ownable, ERC223, SafeMath {
...
-      totalSupply += _initialBalances[i];
-      balances[_initialInvestors[i]] = _initialBalances[i];
-    }
+  /// @param _initialInvestor Address of initial investor
+  /// @param _initialBalance Balance of initial investor
+  function generateTokens(address _initialInvestor, uint _initialBalance) public {
+    require(msg.sender == tokenGenerator || msg.sender == owner);
+
+    totalSupply += _initialBalance;
+    balances[_initialInvestor] = _initialBalance;
```
Comment (EugeneBiazrukau) [inline contracts/Token.sol]: As I can remember, we decided that current logic should call `tokenFallback`method on receiver address. Is this solution still working for us?

File: migrations/2_deploy_token_and_crowdsale.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+const Token = artifacts.require('Token');
+const Crowdsale = artifacts.require('Crowdsale');
+const getConfig = require('../config.js');
+
+module.exports = function (deployer, network, accounts) {
+  let deployedToken;
+  const config = getConfig(accounts);
+  // deployer.deploy(Token, 'MyCHIP', 'CHIP', config.tokenDecimalPlaces, config.options);
```
Comment (EugeneBiazrukau) [inline migrations/2_deploy_token_and_crowdsale.js]: Are we still need this line?

File: test/Token.test.js [hunk 10]
```diff
@@ -38,8 +38,12 @@ contract('Token', function(accounts) {
     it('changes investors balances',async () => {
       const token = await Token.new('chip', 'chip', 0);
       const investorStartBalance = await token.balanceOf(initialAddresses[0]);
-      await token.generateTokens(initialAddresses, initialBalances);
+      await Promise.all([
```
Comment (EugeneBiazrukau) [inline test/Token.test.js]: It will be better to move it to separate function

Comment (EugeneBiazrukau) [reply test/Token.test.js]: Also please add test case for calling method from not permitted addresses

File: test/Crowdsale.test.js (hunk 1 above)
Comment (EugeneBiazrukau) [reply test/Crowdsale.test.js]: Please, look through all tests and remove unused variables

File: test/Crowdsale.test.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+const createNewContract = async (accounts) => {
+  const token = await Token.new('PlayChip', 'CHIP', 0);
+  const crowdsaleContract = await Crowdsale.new(token.address, tokenCost, rate);
+
+  return { token, crowdsaleContract };
+};
+
+contract('PlayChipCrowdsale', (accounts) => {
+  context('After Deploy', () => {
```
Comment (EugeneBiazrukau) [inline test/Crowdsale.test.js:29]: Please, add description and checking values for all constructor arguments. You can do it in one test

**Review by EugeneBiazrukau — CHANGES_REQUESTED** [review]

File: contracts/Token.sol (hunk 8 above)
Comment (niowa) [reply contracts/Token.sol]: @EugeneZombie 
I've realized that logic in `payble` method of Crowdsale contract, but now I see that I need to move tokenFallback logic to this function.

**Review by EugeneBiazrukau — APPROVED** [review]

**Pull request merged**